### PR TITLE
Remote android devices

### PR DESCRIPTION
**Pull request by wordswords**
* Added DeviceAPI::Android.connect and DeviceAPI::Android.disconnect which can be used to connect to remote devices. 
* When a remote device is connected to, it will create a RemoteDevice subclass of Device, which behaves in exactly the same way except it has a different name. If a device is already connected and we are asked to connect to it again, it will throw a catchable custom error called DeviceAlreadyConnected, as well as the standard ADBCommandError when other errors occur. 
* Added tests around this.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,6 +25,18 @@ Try connecting an android device with usb, and run:
...
 
+### Connecting and disconnecting from Remote Devices
+
+You can connect to devices via their IP address and port number. The syntax is:
+
+  DeviceAPI::Android.connect('192.168.0.1:5555')
+
+This should add a device to the already connected devices, which you can query with DeviceAPI::Android.devices. You can disconnect from a device like so:
+
+  DeviceAPI::Android.disconnect('192.168.0.1:5555')
```
Comment (davidbuckhurst) [inline README.md]: I'd like to be able to call disconnect on the device object too

File: lib/device_api/android/adb.rb [hunk 2]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
         shell(serial, cmd)
       end
 
+      def self.check_ip_address(ipaddressandport)
+        unless ipaddressandport =~ /\A(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3}):[0-9]+\Z/
```
Comment (davidbuckhurst) [inline lib/device_api/android/adb.rb]: Should port be optional? Isn't it typically 5555?

File: lib/device_api/android/adb.rb [hunk 3]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
...
+             raise ADBCommandError.new("Invalid IP address and port " + ipaddressandport)
+        end
+      end
+
+      def self.connect(ipaddressandport)
+        check_ip_address(ipaddressandport)
+        cmd = "adb connect " + ipaddressandport
+        result = execute(cmd)
+        if result.stdout.to_s =~ /.*already connected to.*/
+          raise DeviceAlreadyConnectedError.new("Device " + ipaddressandport + " already connected")
```
Comment (davidbuckhurst) [inline lib/device_api/android/adb.rb]: Embed in the string:
"Device #{ipaddressandport} already connected"
(Handles nil and unexpected values better)

File: lib/device_api/android/adb.rb [hunk 4]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
...
+        elsif result.stdout.to_s =~ /.*connected to.*/
+          return
+        else 
+          raise ADBCommandError.new("Unable to adb connect to " + ipaddressandport + " result was: " + result.stdout.inspect)
```
Comment (davidbuckhurst) [inline lib/device_api/android/adb.rb]: Again, embed in one string ... why do you need inspect though? Usually we'd just use that for debug

File: lib/device_api/android/adb.rb [hunk 5]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
...
+        end 
+      end
+
+      def self.disconnect(ipaddressandport)
+        check_ip_address(ipaddressandport)
+        cmd = "adb disconnect " + ipaddressandport
+        result = execute(cmd)
+        unless result.exit == 0
+          raise ADBCommandError.new("Unable to adb disconnect from " + ipaddressandport + " result was: " + result.stdout.inspect)
```
Comment (davidbuckhurst) [inline lib/device_api/android/adb.rb]: Same here

**Comment by davidbuckhurst** [comment]
Good PR -- we should have done this ages ago. Just a few small comments.

File: lib/device_api/android/adb.rb [hunk 6]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
         shell(serial, cmd)
       end
 
+      def self.check_ip_address(ipaddressandport)
+        unless ipaddressandport =~ /\A(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3}):[0-9]+\Z/ 
+             raise ADBCommandError.new("Invalid IP address and port " + ipaddressandport)
+        end
+      end
+
+      def self.connect(ipaddressandport)
```
Comment (jrmhaig) [inline lib/device_api/android/adb.rb]: Similar to the comment above on the `check_ip_address` method I think this should be:

```ruby
def self.connect(ipaddress, port = 5555)
```

File: lib/device_api/android/adb.rb [hunk 7]
```diff
@@ -233,6 +233,34 @@ def self.screencap( serial, args )
...
+        if result.stdout.to_s =~ /.*already connected to.*/
+          raise DeviceAlreadyConnectedError.new("Device " + ipaddressandport + " already connected")
+        elsif result.stdout.to_s =~ /.*connected to.*/
+          return
+        else 
+          raise ADBCommandError.new("Unable to adb connect to " + ipaddressandport + " result was: " + result.stdout.inspect)
+        end 
+      end
+
+      def self.disconnect(ipaddressandport)
```
Comment (jrmhaig) [inline lib/device_api/android/adb.rb]: Like `self.disconnect` the port should be optional, and possibly not required at all.

**Pull request merged**